src/tridiagonal.rs

```diff
@@ -0,0 +1,801 @@
+use super::error::*;
+use super::lapack::*;
+use super::layout::*;
+
+/// Represents a tridiagonal matrix as 3 one-dimensional vectors.
+/// This struct also holds the layout of the raw matrix.
+#[derive(Clone, PartialEq)]
+pub struct TriDiagonal<A: Scalar> {
+    /// layout of raw matrix
+    pub l: MatrixLayout,
+    /// (n-1) sub-diagonal elements of matrix.
+    pub dl: Array1<A>,
+    /// (n) diagonal elements of matrix.
+    pub d: Array1<A>,
+    /// (n-1) super-diagonal elements of matrix.
+    pub du: Array1<A>,
+}
+
+pub trait TridiagIndex {
+    fn to_tuple(&self) -> (i32, i32);
+}
+impl TridiagIndex for [Ix; 2] {
+    fn to_tuple(&self) -> (i32, i32) {
+        (self[0] as i32, self[1] as i32)
+    }
+}
+
+fn debug_bounds_check_tridiag(n: i32, row: i32, col: i32) {
+    if std::cmp::max(row, col) >= n {
+        panic!(
+            "ndarray: index {:?} is out of bounds for array of shape {}",
+            [row, col],
+            n
+        );
+    }
+}
+
+impl<A, I> Index<I> for TriDiagonal<A>
+where
+    A: Scalar,
+    I: TridiagIndex,
+{
+    type Output = A;
+    #[inline]
+    fn index(&self, index: I) -> &A {
+        let (n, _) = self.l.size();
+        let (row, col) = index.to_tuple();
+        debug_bounds_check_tridiag(n, row, col);
+        match row - col {
+            0 => &self.d[row as usize],
+            1 => &self.dl[col as usize],
+            -1 => &self.du[row as usize],
+            _ => panic!(
+                "ndarray-linalg::tridiagonal: index {:?} is not tridiagonal element",
+                [row, col]
+            ),
+        }
+    }
+}
+
+impl<A, I> IndexMut<I> for TriDiagonal<A>
+where
+    A: Scalar,
+    I: TridiagIndex,
+{
+    #[inline]
+    fn index_mut(&mut self, index: I) -> &mut A {
+        let (n, _) = self.l.size();
+        let (row, col) = index.to_tuple();
+        debug_bounds_check_tridiag(n, row, col);
+        match row - col {
+            0 => &mut self.d[row as usize],
+            1 => &mut self.dl[col as usize],
+            -1 => &mut self.du[row as usize],
+            _ => panic!(
+                "ndarray-linalg::tridiagonal: index {:?} is not tridiagonal element",
+                [row, col]
+            ),
+        }
+    }
+}
+
+/// An interface for making a TriDiagonal struct.
+pub trait ToTriDiagonal<A: Scalar> {
+    /// Extract tridiagonal elements and layout of the raw matrix.
+    ///
+    /// If the raw matrix has some non-tridiagonal elements,
+    /// they will be ignored.
+    ///
+    /// The shape of raw matrix should be equal to or larger than (2, 2).
+    fn to_tridiagonal(&self) -> Result<TriDiagonal<A>>;
+}
+
+impl<A, S> ToTriDiagonal<A> for ArrayBase<S, Ix2>
+where
+    A: Scalar + Lapack,
+    S: Data<Elem = A>,
+{
+    fn to_tridiagonal(&self) -> Result<TriDiagonal<A>> {
+        let l = self.square_layout()?;
+        let (n, _) = l.size();
+        if n < 2 {
+            panic!("Cannot make a tridiagonal matrix of shape=(1, 1)!");
```

Comment:
Should not panic here. Please define an Error variant and throw it.

Reply:
I defined a new `LinalgError` named `NotStandardShape` in consideration of versatility.